### PR TITLE
test: skip oomd test on a unified container on a hybrid host

### LINKED ISSUE #20593: test-oomd-util fails in container
**systemd version the issue has been seen with**
 > 248

**Used distribution**
 > Ubuntu 20.04 host (with HWE kernel), Fedora 34 container

**Linux kernel version used** (`uname -a`)

 > `Linux fedora34 5.11.0-27-generic #29~20.04.1-Ubuntu SMP Wed Aug 11 15:58:17 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > amd64 and arm64

**Expected behaviour you didn't see**
 > /usr/lib/systemd/tests/test-oomd-util passing

**Unexpected behaviour you saw**
 > /usr/lib/systemd/tests/test-oomd-util failing like so:

```
[root@fedora34 ~]# /usr/lib/systemd/tests/test-oomd-util 
Error converting 'MemTotal:       32495256 kB trailing' from /oomdgetsysctxtestfM7xTX to uint64_t: Invalid argument
Last pgscan 33 greater than current pgscan 2 for /zupa.slice. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 2 for /zupa.slice. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 2 for /zupa.slice. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Bus n/a: changing state UNSET → OPENING
sd-bus: starting bus by connecting to /run/dbus/system_bus_socket...
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.8 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=StartTransientUnit cookie=3 reply_cookie=0 signature=ssa(sv)a(sa(sv)) error-name=n/a error-message=n/a
Got message type=method_return sender=:1.4 destination=:1.8 path=n/a interface=n/a member=n/a cookie=114 reply_cookie=3 signature=o error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.DBus destination=:1.8 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=4294967295 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.8 path=n/a interface=n/a member=n/a cookie=4294967295 reply_cookie=2 signature= error-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.systemd1',path='/org/freedesktop/systemd1',interface='org.freedesktop.systemd1.Manager',member='JobRemoved' successfully installed.
Got message type=signal sender=:1.4 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=119 reply_cookie=0 signature=uoss error-name=n/a error-message=n/a
Got result done/Success for job test-oomd-util-8d268a3c9d766cf0.scope
Bus n/a: changing state RUNNING → CLOSED
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
oomd attempting to kill 157 with KILL
oomd attempting to kill 158 with KILL
oomd attempting to kill 159 with KILL
oomd attempting to kill 160 with KILL
Error getting memory.current from /system.slice/test-oomd-util-8d268a3c9d766cf0.scope: No data available
Assertion 'oomd_cgroup_context_acquire(cgroup, &ctx) == 0' failed at src/oom/test-oomd-util.c:117, function test_oomd_cgroup_context_acquire_and_insert(). Aborting.
Aborted (core dumped)
```

**Steps to reproduce the problem**

1. Get yourself an Ubuntu 20.04 system
2. lxd init, accept all the defaults
3. lxc launch images:fedora/34 -c security.nesting=true fedora34
4. Install systemd-tests inside the container
5. Run /usr/lib/systemd/tests/test-oomd-util

**Additional program output to the terminal or log subsystem illustrating the issue**

-  The "-c security.nesting=true" when launching the container seems to be necessary to get systemd in the container to set up a unified hierarchy as the failing bits of test-oomd-util don't run for a hybrid hierarchy
- Although 20.04 uses a hybrid hierarchy by default, changing the kernel command line to force a unified hierarchy does not make the tests pass (so it's not a "hybrid outside" / "unified inside" sort of thing) 
- I tried with a Fedora 34 vm as a host and didn't see the failure. That has a 5.13 kernel though, which might be relevant?

### LINKED ISSUE #20655: test-oomd-util keeps failing in Semaphore CI
**systemd version the issue has been seen with**
 > latest master

**Used distribution**
 > Debian Bullseye (LXC in Semaphore CI)

In recently pushed PRs the Semaphore CI job started failing in `test-oomd-util`:

```
The output from the failed tests:

 402/1230 test-oomd-util                                                                              FAIL           0.01s (killed by signal 6 SIGABRT)

--- command ---
07:11:29 SYSTEMD_LANGUAGE_FALLBACK_MAP='/tmp/autopkgtest-lxc.zbkaaagv/downtmp/build.qgM/src/src/locale/language-fallback-map' SYSTEMD_KBD_MODEL_MAP='/tmp/autopkgtest-lxc.zbkaaagv/downtmp/build.qgM/src/src/locale/kbd-model-map' PATH='/tmp/autopkgtest-lxc.zbkaaagv/downtmp/build.qgM/src/build-deb:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin' /tmp/autopkgtest-lxc.zbkaaagv/downtmp/build.qgM/src/build-deb/test-oomd-util
--- stderr ---
Error converting 'MemTotal:       32495256 kB trailing' from /oomdgetsysctxtestFafYLd to uint64_t: Invalid argument
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 2 for /zupa.slice. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 2 for /zupa.slice. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 2 for /zupa.slice. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 2 for /zupa.slice. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Last pgscan 33 greater than current pgscan 1 for /herp.slice/derp.scope. Using last pgscan of zero.
Bus n/a: changing state UNSET → OPENING
sd-bus: starting bus by connecting to /run/dbus/system_bus_socket...
Bus n/a: changing state OPENING → AUTHENTICATING
Bus n/a: changing state AUTHENTICATING → HELLO
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=Hello cookie=1 reply_cookie=0 signature=n/a error-name=n/a error-message=n/a
Sent message type=method_call sender=n/a destination=org.freedesktop.DBus path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=AddMatch cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.12 path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=n/a error-message=n/a
Bus n/a: changing state HELLO → RUNNING
Sent message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=StartTransientUnit cookie=3 reply_cookie=0 signature=ssa(sv)a(sa(sv)) error-name=n/a error-message=n/a
Got message type=method_return sender=:1.1 destination=:1.12 path=n/a interface=n/a member=n/a cookie=877 reply_cookie=3 signature=o error-name=n/a error-message=n/a
Got message type=signal sender=org.freedesktop.DBus destination=:1.12 path=/org/freedesktop/DBus interface=org.freedesktop.DBus member=NameAcquired cookie=2 reply_cookie=0 signature=s error-name=n/a error-message=n/a
Got message type=method_return sender=org.freedesktop.DBus destination=:1.12 path=n/a interface=n/a member=n/a cookie=3 reply_cookie=2 signature=n/a error-name=n/a error-message=n/a
Match type='signal',sender='org.freedesktop.systemd1',path='/org/freedesktop/systemd1',interface='org.freedesktop.systemd1.Manager',member='JobRemoved' successfully installed.
Got message type=signal sender=:1.1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=882 reply_cookie=0 signature=uoss error-name=n/a error-message=n/a
Got result done/Success for job test-oomd-util-f69769737b554529.scope
Bus n/a: changing state RUNNING → CLOSED
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
test-oomd-util: Cannot set user xattrs, skipping tests.
Error getting memory.current from /system.slice/test-oomd-util-f69769737b554529.scope: No data available
Assertion 'oomd_cgroup_context_acquire(cgroup, &ctx) == 0' failed at src/oom/test-oomd-util.c:117, function test_oomd_cgroup_context_acquire_and_insert(). Aborting.
```

Seen in:
  * #20351: https://the-real-systemd.semaphoreci.com/jobs/859d6f4e-61cb-4623-9295-175827101724
  * #20226: https://the-real-systemd.semaphoreci.com/jobs/109727ae-ea20-4293-b3e2-66de2654abab

/cc @anitazha

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #20593
Fixes #20655.

### CODE REVIEW

**Comment by yuwata** [comment]
(I have no idea what we should do in oomd on such a case.)

**Comment by yuwata** [comment]
cc @anitazha and @DaanDeMeyer

**Comment by DaanDeMeyer** [comment]
Slightly out of my league as well. Let's wait for anita.

**Comment by anitazha** [comment]
I need to check in more detail when I'm not AFK but memory.current is provided when the memory controller is enabled. So I think the correct message would be something like "cgroup2 memory controller is not enabled". And the actual daemon should also include a condition check for the memory controller.

**Comment by yuwata** [comment]
Ah, checking the existence of memory controller is better. Thanks. Will update.

**Comment by yuwata** [comment]
@anitazha Updated. PTAL.

**Comment by bluca** [comment]
test-oomd still unhappy

**Comment by yuwata** [comment]
Still not sure this is a correct way, but at least, now semaphoreci is green. PTAL.

**Comment by yuwata** [comment]
What's the difference between `cg_kernel_controllers()` and `cg_mask_supported()`??

File: src/oom/oomd.c [hunk 1]
```diff
@@ -153,6 +155,20 @@ static int run(int argc, char *argv[]) {
         if (r == 0)
                 return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Requires the unified cgroups hierarchy");
 
+        r = cg_kernel_controllers(&controllers);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get supported cgroup controllers: %m");
+
+        if (!set_contains(controllers, "memory"))
+                return log_error_errno(SYNTHETIC_ERRNO(EOPNOTSUPP), "Requires the cgroup memory controller.");
+
```
Comment (anitazha) [inline src/oom/oomd.c]: This is not needed since we have `cg_mask_supported` below

File: src/oom/test-oomd-util.c [hunk 2]
```diff
@@ -102,6 +104,16 @@ static void test_oomd_cgroup_context_acquire_and_insert(void) {
         if (cg_all_unified() <= 0)
                 return (void) log_tests_skipped("cgroups are not running in unified mode");
 
+        assert_se(cg_kernel_controllers(&controllers) >= 0);
+
+        if (!set_contains(controllers, "memory"))
+                return (void) log_tests_skipped("cgroup memory controller is not available");
+
```
Comment (anitazha) [inline src/oom/test-oomd-util.c]: This is not needed since we have `cg_mask_supported` below

**Review by anitazha — CHANGES_REQUESTED** [review]
> What's the difference between `cg_kernel_controllers()` and `cg_mask_supported()`??

It looks like `cg_kernel_controllers()` parses /proc/cgroup which lists all the controllers (cgroup v1 and v2) compiled in the kernel (including information about how many cgroups are using them and whether they were disabled via boot parameter, etc.). But `cg_mask_supported()` will figure out which hierarchy systemd is using and list the available controllers based on the current hierarchy. For unified (cgroup v2) this will be /sys/fs/cgroups/cgroup.controllers.

Thanks for fixing this Yu the current approach lgtm. Should be good to merge after removing the calls to `cg_kernel_controllers()`

**Comment by yuwata** [comment]
@anitazha Thank you for your review and the explanation. `cg_kernel_controllers()` are dropped. Upgrading the green label.

**Review by bluca — APPROVED** [review]
LGTM, thanks for fixing this, we can go back to seeing all green ticks yay

**Pull request merged**